### DISCUSSION TITLE
Permission Denied Error whenever Create/Edit/Save Files using Atom - Windows Subsystem for Linux - WSL 1

### DISCUSSION THREAD

**Discussion by hyeom-wisc** [idaholab/moose#16459] [Q&A Getting Started] [answered]
Hello,
I have worked on tutorial #1 and moved forward by your help. Thanks a lot.

One thing I noticed was that there has been "Permission denied" error whenever I created, edited, and saved input files and source codes using ATOM. So far, I have provided the permission to each file using "chmod +rw" command manually. As you expected, it is time consuming. Is there any way to give the permission to files created/modified by ATOM easily?

Sincerely,
Sung

**Comment by cticenhour** [comment]
The first thing that comes to my mind - do you launch atom using `sudo`? If opened as root, even once, a given file (or project directory) could have screwed up permissions like this. Changing a single file back to the proper permissions wouldn't fix an entire directory tree that is screwed up in this way.

Alternatively, do you install packages using the `apm install` command on the command line? This is mentioned in the [Atom Editor Setup page](https://mooseframework.inl.gov/help/development/Atom_Editor.html) on the MOOSE website. I am wondering if you accidentally used `sudo apm install` which might have screwed up permissions in your Atom installation, and is spilling down to what you edit using it. Though that's a bit more far-fetched, perhaps.

**Reply by cticenhour** [reply]
One more thing to add in general....In normal Atom usage, you should never need to mess with permissions on files that you have created or downloaded. So this is definitely a-typical.

**Reply by hyeom-wisc** [reply]
Casey,

Thank you for the quick response. Only thing I did once I installed ATOM is to create files and save in directories of interest. Honestly, I cannot understand the Atom Editor Setup page. I was stuck even in the first step. Do I need to do everything shown on the page? 
Sung

**Reply by cticenhour** [reply]
Only if you want to - plenty of developers use Atom without language and development plug-ins, which is what that page is discussing. Feel free to ask a question about that as well if you are confused. 

My next question is then: how do you launch Atom? Graphically (by clicking on it) or through the Terminal?

**Reply by hyeom-wisc** [reply]
Thank you. It seems I need some help on this. I just downloaded/installed Atom from the website and click the icon to run Atom in Window. I created files (input files, source codes) using Atom and saved in directories. And then I ran the files in Moose framework using Ubuntu terminal. This time I saw the error messages and changed the permission one by one. I don't think this is typical.
Sung

**Reply by cticenhour** [reply]
Do you store your source code and input files within a MOOSE app that you created? If so, could you give the output of `ls -la` when you `cd` to that location in your Terminal? I am wondering now if it's not Atom at all, but how that application was made in the first place.

**Reply by hyeom-wisc** [reply]
Please see the output in the terminal when I typed ls -la in the MOOSE app. 
(moose) hyeom@ERB914-7920:~$ ls -la
total 90900
drwxr-xr-x 1 hyeom hyeom     4096 Dec 10 15:33 .
drwxr-xr-x 1 root  root      4096 Nov 16 16:00 ..
-rw------- 1 hyeom hyeom    14453 Dec 10 16:46 .bash_history
-rw-r--r-- 1 hyeom hyeom      220 Nov 16 16:00 .bash_logout
-rw-r--r-- 1 hyeom hyeom     4331 Nov 27 10:57 .bashrc
drwxr-xr-x 1 hyeom hyeom     4096 Nov 27 11:13 .cache
drwxr-xr-x 1 hyeom hyeom     4096 Nov 16 16:16 .conda
-rw-r--r-- 1 hyeom hyeom       52 Nov 23 10:35 .condarc
drwxr-xr-x 1 hyeom hyeom     4096 Nov 27 11:13 .config
-rw-r--r-- 1 hyeom hyeom       26 Nov 24 12:23 .gitconfig
drwxr-xr-x 1 hyeom hyeom     4096 Nov 16 16:02 .landscape
drwxr-xr-x 1 hyeom hyeom     4096 Dec  7 10:26 .local
-rw-r--r-- 1 hyeom hyeom        0 Dec 11 09:16 .motd_shown
-rw-r--r-- 1 hyeom hyeom      807 Nov 16 16:00 .profile
drwx------ 1 hyeom hyeom     4096 Nov 28 15:14 .ssh
-rw-r--r-- 1 hyeom hyeom        0 Nov 16 16:04 .sudo_as_admin_successful
-rw-r--r-- 1 hyeom hyeom      478 Nov 16 16:36 .zshrc
-rw-r--r-- 1 hyeom hyeom 93052469 Nov 16 16:31 Miniconda3-latest-Linux-x86_64.sh
drwxr-xr-x 1 hyeom hyeom     4096 Nov 16 16:16 miniconda3
drwxr-xr-x 1 hyeom hyeom     4096 Dec  7 09:39 projects

**Reply by cticenhour** [reply]
Thank you for the output. Instead of the home directory (`~`), I wanted you to run this in your projects directory (or even in the application you've been developing within projects). so `cd ~/projects` and then `ls -la`. But I do see something interesting here in your home directory. You do have a `.sudo_as_admin_successful` file dated on November 16th, around the same time you installed conda. This file is created after using sudo so that the system knows not to show a bunch of warnings about the dangers of sudo to you after you've used the command once. I am still wondering if this is a general permissions issue that you only see when using Atom. Could you give me that second set of Terminal output in your projects directory?

**Reply by cticenhour** [reply]
One more question - did you use `sudo` to install Miniconda?

**Reply by hyeom-wisc** [reply]
Please check the outputs when I typed ls -la in projects folder and my application folder. I don't remember I used sudo for Miniconda.

(moose) hyeom@ERB914-7920:~/projects$ ls -la
total 0
drwxr-xr-x 1 hyeom hyeom 4096 Dec  7 09:39 .
drwxr-xr-x 1 hyeom hyeom 4096 Dec 11 10:08 ..
drwxr-xr-x 1 hyeom hyeom 4096 Dec  7 10:05 babbler
drwxr-xr-x 1 hyeom hyeom 4096 Dec 10 15:59 babbler2
drwxr-xr-x 1 hyeom hyeom 4096 Nov 24 11:58 hyeom
drwxr-xr-x 1 hyeom hyeom 4096 Nov 16 16:47 moose
drwxr-xr-x 1 hyeom hyeom 4096 Nov 28 14:03 practice
drwxr-xr-x 1 hyeom hyeom 4096 Nov 17 16:49 your_app_name

(moose) hyeom@ERB914-7920:~/projects/babbler2$ ls -la
total 476
drwxr-xr-x 1 hyeom hyeom   4096 Dec 10 15:59 .
drwxr-xr-x 1 hyeom hyeom   4096 Dec  7 09:39 ..
-rw-r--r-- 1 hyeom hyeom    738 Dec  7 09:39 .clang-format
drwxr-xr-x 1 hyeom hyeom   4096 Dec 11 10:21 .git
-rw-r--r-- 1 hyeom hyeom   3266 Dec  7 09:39 .gitignore
drwxr-xr-x 1 hyeom hyeom   4096 Dec  7 09:45 .libs
-rw-r--r-- 1 hyeom hyeom   7848 Dec  7 09:46 .previous_test_results.json
-rw-r--r-- 1 hyeom hyeom  26530 Dec  7 09:39 LICENSE
-rw-r--r-- 1 hyeom hyeom   2061 Dec  7 09:39 Makefile
-rw-r--r-- 1 hyeom hyeom    196 Dec  7 09:56 README.md
-rwxr-xr-x 1 hyeom hyeom 435696 Dec 10 15:59 babbler2-opt
drwxr-xr-x 1 hyeom hyeom   4096 Dec  7 12:05 build
drwxr-xr-x 1 hyeom hyeom   4096 Dec  7 09:39 doc
drwxr-xr-x 1 hyeom hyeom   4096 Dec  7 11:59 include
drwxr-xr-x 1 hyeom hyeom   4096 Dec 10 15:59 lib
drwxr-xr-x 1 hyeom hyeom   4096 Dec  7 12:22 problems
-rwxr-xr-x 1 hyeom hyeom    471 Dec  7 09:39 run_tests
drwxr-xr-x 1 hyeom hyeom   4096 Dec  7 09:39 scripts
drwxr-xr-x 1 hyeom hyeom   4096 Dec 10 15:56 src
drwxr-xr-x 1 hyeom hyeom   4096 Dec  7 09:45 test
-rw-r--r-- 1 hyeom hyeom     87 Dec  7 09:39 testroot
drwxr-xr-x 1 hyeom hyeom   4096 Dec  7 09:39 unit

**Comment by GiudGiud** [comment]
If you need a quick fix I reckon you should un-install and re-install Atom. 
Avoid using sudo in that process.

**Reply by hyeom-wisc** [reply]
I have uninstalled Atom and reinstalled it. I tried to set up Atom with following steps shown in the instruction. https://mooseframework.inl.gov/help/development/Atom_Editor.html
However, there was error in the section for Source navigation with rtags. Anyway, I made a new input file using Atom and I still see the same permission denied error in the terminal.

**Reply by dschwen** [reply]
Yeah, installing Atom with sudo cannot cause this issue. _Where_ are you saving your input file to? This must be a file system permission issue. You've shown us the listing of your home directory, but are you actually saving to that location?

**Reply by hyeom-wisc** [reply]
I have saved files created in Atom under the folder as follows:
C:\Users\hyeom\AppData\Local\Packages\CanonicalGroupLimited.Ubuntu20.04onWindows_79rhkp1fndgsc\LocalState\rootfs\home\hyeom\projects

I had the same permission issue when I ran git commands (e.g., git status). That time I provided the permission manually again.

**Reply by cticenhour** [reply]
Are you editing these files in Windows and then running them in your WSL Ubuntu instance?

**Reply by hyeom-wisc** [reply]
I didn't edit the files in Windows file explorer (I check the directory, yes) but edited through Ubuntu Terminal and Atom.

**Reply by cticenhour** [reply]
By editing I was referring to Atom - do you have Atom installed in Windows, launch it in Windows, and are using it to edit inside of your projects directory? This could be causing the permissions issue you describe. Windows Subsystem for Linux generally doesn't handle that well.

**Reply by hyeom-wisc** [reply]
Yes. I installed and ran Atom in Windows. I found and opened the files in the Windows directory having applications. What is the correct way to access or save the files through Atom? When I created a new file using Atom, I needed to convert the file using dos2unix command.

**Reply by cticenhour** [reply]
It is not recommended to use Atom in this way on Windows with WSL Ubuntu, because of the issues you describe. Windows does a poor job handling permissions and translating between the two OS instances (as you've discovered). I would recommend that you don't use Atom for your WSL development, and use something like `vim`, `emacs`, or `nano`, which you can install natively inside of your Ubuntu installation. These do have a tiny learning curve and involve working in a Terminal over having a graphical environment.

**Reply by hyeom-wisc** [reply]
Thank you so much for the recommendation. Can you suggest any link or training website for me to start the text editors?

**Reply by cticenhour** [reply]
Of course, you can continue to use Atom and perform the conversions and permissions changes as you're doing. But, there is currently no way to guarantee a "clean" experience using Atom in Windows with Ubuntu in WSL.

**Reply by cticenhour** [reply]
Here is a guided tutorial for vim (which is what I use when I'm not using Atom on Mac or Linux) https://www.openvim.com/. More tutorials for vim, emacs, or nano can be easily found via a Google search. 

EDIT: Here's a written one for vim that also seems quite good https://danielmiessler.com/study/vim/

**Reply by GiudGiud** [reply]
You could also install atom on WSL. It should work

**Reply by hyeom-wisc** [reply]
Thank you all. Let me try using vim as a text editor. It seems using Atom on Windows with WSL Ubuntu is not easy.

**Reply by GiudGiud** [reply]
I mean Atom installed in WSL, ran from WSL
I mean maybe you'll like VIM, but it's pretty minimal.

**Reply by hyeom-wisc** [reply]
All,
After I opened the input files using Vim or Notepad ++ in Windows 10 and saved them, I lost the permission to the files in Ubuntu. This is the same thing when I used Atom in Windows 10. Was my approach wrong?

**Reply by GiudGiud** [reply]
Please try opening the input files from a text editor in Ubuntu / WSL. Any text editor opened on the windows side will change the permissions.

**Reply by cticenhour** [reply]
Also I do want to suggest you give a try to @dschwen's suggestion below as well. I'll try to remember the steps without having Windows in front of me (but feel free to correct me Daniel!). To find the virtual network share he mentions, open a new Atom window, click "Add project folder" in the drop down menus. In the Open Folder prompt that Atom provides, look on the left hand side of the window (you might have to scroll down the menu panel), select "Network", then "Ubuntu-##.##", then you should have your Linux file system and can navigate to your `projects/moose` directory. That would allow a Windows workflow for your code writing, while you compile and run inside of WSL.

**Reply by hyeom-wisc** [reply]
All,
Thanks. I have decided to use vim to edit text files through WSL. There is no permission issue now.

**Comment by cticenhour** [comment]
For those who come across this later, the answer to this issue was that Atom in Windows was being used to edit files within Windows Subsystem for Linux. Windows does not do a great job handling the transition between the two OS environments and often causes issues related to corrupted files, file permissions, or other incompatibilities. Therefore, it is recommended that a Linux native editor is used. We recommend `vim` and `emacs` since many members of the MOOSE development ecosystem use these, as well as other lightweight editors like `nano`. Tutorials for vim can be found in the threads in this discussion, as well as via a Google search.

**Reply by dschwen** [reply]
No @cticenhour , see below.

**Reply by cticenhour** [reply]
I have no knowledge of WSL2, so thanks for the additional information @dschwen :) 

This answer would still apply for WSL1 users though.

**Comment by dschwen** [comment, marked as answer]
The solution is to use **WSL2** and access files through a virtual network share located at `\\wsl$\distribution\path`. I have gone through the effort of enabling autocompletion with native Atom on Windows with WSL2 based MOOSE installations.

**Reply by dschwen** [reply]
![image](https://user-images.githubusercontent.com/202302/101940380-a0e4f680-3ba3-11eb-9ae8-d2f5507abfeb.png)

**Reply by dschwen** [reply]
From our FY 2020 report:

> WSL2 promised better performance and we investigated building MOOSE and MOOSE-based applications on this improved system. Besides having to build the MPI mpich2 from scratch, we were able to follow the Linux build instructions very closely and ended up with fully featured MOOSE app executables.
> To improve the user experience, the MOOSE integration for autocompletion[2] in the Atom editor[3] was improved to work on Windows allowing the native Windows version of Atom to be used to edit files inside the WSL2 container. WSL2 files are made available in Windows through a virtual network share located at `\\wsl$\distribution\path`. The moose-autocomplete plugin for Atom obtains a description of valid application syntax by running a MOOSE app using the `--json` commandline switch. This causes the app to output JSON data describing all valid parameters for all available objects for the given app — rather than statically supplying a syntax description which would quickly go out of date as soon as objects in a MOOSE-based application are added or modified.
The MOOSE-autocomplete plugin detects whenever a file is located within a WSL2 container and uses the Windows `wsl` command to discover and execute a MOOSE app executable within the WSL2 container. The entire process is seamless to the user. Figure 6.2 shows a screenshot of a native Atom editor being used to edit a MOOSE inputfile within a WSL2 container with working autocompletion. We believe WSL2 to be a suitable avenue for making MOOSE available on Windows systems as it very closely maps to our current Linux based testing and development infrastructure.

**Reply by cticenhour** [reply]
We really should integrate this into our current documentation for Windows 10, especially since users can still choose WSL1 or WSL2.
